### PR TITLE
Have Section and MitoSection have value semantics for comparison

### PR DESCRIPTION
**Pull request by mgeplf**
* two sections are the same when they have the same
  points/diameters/etc, not if they belong to the same instance of the
  same morphology

### CODE REVIEW

**Comment by mgeplf** [comment]
@eleftherioszisis: do you think that value semantics are what is expected for `==`; or should I have another method for comparison: `has_same_structure` or something?

**Comment by eleftherioszisis** [comment]
I think the way you have it is fine. Shouldn't we also check that the topology is identical too, i.e. that they have the same parent/children?

**Comment by eleftherioszisis** [comment]
Maybe have the convention that `==` would require everything to be identical, topology included, while `has_same_values` only the values?

**Comment by mgeplf** [comment]
> `topology included`

I'm not sure that the number of children is intrinsic to the 'sameness' of the section; if it is, then would it need to be recursively true (ie: that the children are the same?) or just that the number matches?

I don't have a clear feeling of which one is 'right', if one can even say there is a 'right'.

**Comment by eleftherioszisis** [comment]
Hmm, I was thinking in terms of ids, but I think you are right. If we include the topology we will more all less need to traverse the tree to ensure that the children and the parent are identical too. I can see two options:
1. delete the equality operator so that it cannot be used and leave the `has_same_values` method.
2. Limit the equality operator to check the ids of parent and children. I believe it is not very likely there will be two sections with the same id and differing values, right?

**Comment by mgeplf** [comment]
I think I'll rename it `has_same_values`, and remove the equality (like you said).

File: include/morphio/section_base.h [hunk 1]
```diff
@@ -98,6 +83,16 @@ SectionBase<T>::SectionBase(uint32_t id, const std::shared_ptr<Property::Propert
...
+template <typename T>
+inline bool SectionBase<T>::operator==(const SectionBase& other) const noexcept {
+    return other.id_ == id_ && other.properties_ == properties_;
+}
+
+template <typename T>
+inline bool SectionBase<T>::operator!=(const SectionBase& other) const noexcept {
+    return !(*this == other);
+}
+
```
Comment (eleftherioszisis) [inline include/morphio/section_base.h:95]: Just for our sanity, wouldn't it be better to also declare them on the binds side?
```
pybind11_class.def(pybind11::self == pybind11::self);
pybind11_class.def(pybind11::self != pybind11::self);
```
For both mut and immut?

File: src/section.cpp [hunk 2]
```diff
@@ -55,6 +55,11 @@ range<const floatType> Section::perimeters() const {
     return get<Property::Perimeter>();
 }
 
+bool Section::hasSameShape(const Section& other) const noexcept {
```
Comment (eleftherioszisis) [inline src/section.cpp:58]: `hasSameShape` is a bit arbitrary (it usually refers to dimensions), isn't it? Wouldn't a name such as `hasSameData` or `hasSameValues` be a little more informative?

Comment (mgeplf) [reply src/section.cpp:58]: `Shape` I think is closest to what I mean; `Data`/`Value` is too generic; the diameters, etc, all define the shape of the section, but the child count doesn, imo.

File: include/morphio/section_base.h (hunk 1 above)
Comment (mgeplf) [reply include/morphio/section_base.h:95]: python is already kinda doing this, from what I can tell.  If we need it, let's do it in another PR

Comment (eleftherioszisis) [reply include/morphio/section_base.h:95]: okk

File: src/section.cpp (hunk 2 above)
Comment (eleftherioszisis) [reply src/section.cpp:58]: `hasSameGeometry` maybe?

**Review by eleftherioszisis — APPROVED** [review]

**Pull request merged**